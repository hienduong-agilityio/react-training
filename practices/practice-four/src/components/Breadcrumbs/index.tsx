// Libraries
import { Link } from 'react-router-dom';
import { memo } from 'react';

// SVG
import homeIcon from '@public/images/homeIcon.svg';
import arrowRight from '@public/images/arrowRightIcon.svg';

// Constants
import { ROUTE } from '@/constants';

export type BreadcrumbItem = {
  // id: The breadcrumb Items id
  id?: number;
  // label: The label to display for the breadcrumb item
  label?: string;
  // href: Optional URL for the breadcrumb item
  href?: string;
};

interface IBreadcrumbsProps {
  // items: Array of breadcrumb items
  items?: BreadcrumbItem[];
}

/**
 * Breadcrumbs component
 *
 * @returns {JSX.Element} Breadcrumbs element
 */
const Breadcrumbs = ({ items = [{ label: 'Home', href: ROUTE.ROOT }] }: IBreadcrumbsProps): JSX.Element => {
  return (
    <nav aria-label='breadcrumb'>
      <ul className='flex items-center space-x-1 md:space-x-2 rtl:space-x-reverse'>
        <li className='inline-flex items-center'>
          <Link to={ROUTE.ROOT} className='flex items-center text-sm font-medium text-gray-400 hover:text-primary-500'>
            <img src={homeIcon} alt='home' className='w-4 h-4 me-2.5 p-0' />
          </Link>
        </li>
        {items.map((item) => (
          <li key={item.label || item.id} className='inline-flex items-center'>
            <img src={arrowRight} alt='Arrow Right' className='rtl:rotate-180 w-6 h-6 text-gray-400 mx-1' />
            {item.href ? (
              <Link to={item.href} className='ms-1 text-sm font-medium text-gray-400 hover:text-primary-500 md:ms-2'>
                {item.label}
              </Link>
            ) : (
              <span className='ms-1 text-sm font-medium text-gray-400 md:ms-2'>{item.label}</span>
            )}
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default memo(Breadcrumbs);
