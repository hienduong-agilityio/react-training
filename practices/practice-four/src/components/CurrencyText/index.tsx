export interface ICurrencyTextProps extends React.HTMLAttributes<HTMLParagraphElement> {
  // Currency: Optional estimated cost of the project, displayed as a financial estimate.
  currency?: number;
}

/**
 * CurrencyText component
 *
 * @returns {JSX.Element} - CurrencyItem element
 */
export const CurrencyText = ({ currency = 0, className = '' }: ICurrencyTextProps): JSX.Element => {
  return <p className={`${className}`}>{currency ? `US$ ${currency}k` : '-'}</p>;
};
