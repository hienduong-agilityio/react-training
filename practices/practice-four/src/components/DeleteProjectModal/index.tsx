// Components
import Modal from '@/components/common/Modal';
import Button from '@/components/common/Button';
import Spinner from '@/components/common/Spinner';

// Enums
import { BUTTON_COLORS, BUTTON_VARIANTS } from '@/enums/theme';

// Hooks
import { useProject } from '@/hooks/useProject';

// Store
import { ToastStore } from '@/stores/toastStore';

// Types
import type { IProjectItemProps } from '@/interfaces';

export interface IDeleteProjectModalProps {
  // projectId: The project id.
  projectId: string;
  // isModalOpen: A boolean indicating whether the modal is open.
  isModalOpen: boolean;
  // onCloseModal: A function to be called when the modal is requested to be closed.
  onCloseModal: () => void;
}

/**
 * The custom DeleteProjectModal component
 *
 * @returns {JSX.Element} - The DeleteProjectModal element
 */
export const DeleteProjectModal = ({
  projectId,
  isModalOpen = false,
  onCloseModal = () => {}
}: IDeleteProjectModalProps): JSX.Element => {
  const project = useProject({ id: projectId || '' });
  const { showToast } = ToastStore();

  const {
    mutate: { deleteProject },
    data: projectDetail,
    isMutating
  } = project;

  const projectName = (projectDetail as IProjectItemProps)?.projectName || 'this project';

  const handleDeleteProjectSuccess = () => {
    showToast('Project Deleted successfully', 'success');
    onCloseModal();
  };

  const handleDeleteProjectError = () => {
    showToast('Failed to deleted project', 'error');
    onCloseModal();
  };

  const handleDelete = () => {
    if (projectId) {
      deleteProject(projectId, {
        onSuccess: handleDeleteProjectSuccess,
        onError: handleDeleteProjectError
      });
    }
  };

  return (
    <Modal isOpen={isModalOpen} onClose={onCloseModal}>
      <div className='bg-light rounded-t-xl p-4 w-[440px]'>
        <p className='text-lg font-semibold text-gray-900 mb-3'>Delete Project</p>
        <p className='leading-6 text-gray-600'>
          Are you sure you want to delete <strong>{projectName}</strong>? If you delete, it will be permanently lost.
        </p>
      </div>
      <div className='bg-light rounded-b-xl flex justify-end gap-4 border-t-2 py-3'>
        <Button onClick={onCloseModal} variant={BUTTON_VARIANTS.OUTLINED}>
          Cancel
        </Button>
        <Button
          onClick={handleDelete}
          disabled={isMutating}
          variant={BUTTON_VARIANTS.CONTAINED}
          color={BUTTON_COLORS.DANGER}
          customClasses='mr-5'
        >
          {isMutating ? <Spinner /> : 'Delete'}
        </Button>
      </div>
    </Modal>
  );
};
