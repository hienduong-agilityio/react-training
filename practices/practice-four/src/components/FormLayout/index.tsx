interface FormLayoutProps {
  title: string;
  subtitle?: string;
  children: React.ReactNode;
}

export const FormLayout = ({ title, subtitle, children }: FormLayoutProps) => (
  <section>
    <div className='flex rounded-t-xl items-center bg-white justify-between p-4'>
      <p className='text-xl font-semibold text-gray-900'>{title}</p>
    </div>
    {subtitle && <p className='text-gray-600 mt-1 mb-4'>{subtitle}</p>}
    {children}
  </section>
);
