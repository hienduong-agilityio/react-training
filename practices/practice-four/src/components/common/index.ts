export { default as Avatar } from './Avatar';
export { default as Button } from './Button';
export { default as InputField } from './InputField';
export { default as Modal } from './Modal';
export { default as Pagination } from './Pagination';
export { default as SelectField } from './SelectField';
export * from './SkeletonLoader';
export { default as Spinner } from './Spinner';
export { default as StatusLabel } from './StatusLabel';
export { default as Tag } from './Tag';
export { default as Toast } from './ToastMessage';
