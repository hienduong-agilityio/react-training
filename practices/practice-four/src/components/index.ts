export { default as Breadcrumbs } from './Breadcrumbs';
export * from './CurrencyText';
export * from './DeleteProjectModal';
export * from './ErrorBoundary';
export * from './ErrorDisplay';
export * from './ProjectDetailSkeleton';
export { default as ProjectFilterDropdown } from './ProjectFilterDropdown';
export * from './ProjectFilterDropdown';
export { default as ProjectItem } from './ProjectItem';
export * from './ProjectItem';
export * from './ProjectOptionsDropdown';
export * from './ProjectTable';
export { default as ResourceTag } from './ResourcesTag';
export { default as SearchBox } from './SearchBox';
export * from './SuspenseLoader';
export { default as Timeline } from './Timeline';
export * from './ToastContainer';
export * from './FormLayout';
export * from './ProjectFormFields';
export * from './SubmitForm';

// Common
export * from './common';
