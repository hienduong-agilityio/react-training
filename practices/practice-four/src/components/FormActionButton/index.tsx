// Libraries
import { useFormStatus } from 'react-dom';

// Components
import { Button, Spinner } from '@/components';

// Enums
import { BUTTON_COLORS, BUTTON_SIZES, BUTTON_VARIANTS } from '@/enums';

interface IFormActionButtonProps {
  children: React.ReactNode;
  type?: 'submit' | 'button';
  color?: BUTTON_COLORS;
  variant?: BUTTON_VARIANTS;
  size?: BUTTON_SIZES;
  disabledOnPending?: boolean;
  showSpinner?: boolean;
  onClick?: () => void;
  customClass?: string;
}

export const FormActionButton = ({
  children,
  type = 'submit',
  variant = BUTTON_VARIANTS.CONTAINED,
  size = BUTTON_SIZES.LARGE,
  color = BUTTON_COLORS.PRIMARY,
  disabledOnPending = true,
  showSpinner = true,
  customClass = '',
  onClick
}: IFormActionButtonProps) => {
  const { pending } = useFormStatus();

  return (
    <Button
      type={type}
      disabled={disabledOnPending && pending}
      variant={variant}
      size={size}
      color={color}
      onClick={onClick}
      customClasses={customClass}
    >
      {pending && showSpinner ? <Spinner /> : children}
    </Button>
  );
};
