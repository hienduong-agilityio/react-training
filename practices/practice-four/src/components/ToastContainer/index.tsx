// Stores
import { ToastStore } from '@/stores';

// Components
import { Toast } from '@/components';

export const ToastContainer = () => {
  const { message, type, isVisible, undoEnabled, timeoutDuration, hideToast } = ToastStore();

  return (
    <>
      {isVisible && (
        <div className='fixed top-9 left-1/2 transform -translate-x-1/2'>
          <Toast type={type} undoEnabled={undoEnabled} timeoutDuration={timeoutDuration} onClose={hideToast}>
            {message}
          </Toast>
        </div>
      )}
    </>
  );
};
