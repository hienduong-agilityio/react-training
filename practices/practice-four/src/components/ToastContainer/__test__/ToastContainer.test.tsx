// Libraries
import { render, screen } from '@testing-library/react';

// Stores
import { ToastStore } from '@/stores';

// Components
import { ToastContainer } from '@/components';

jest.mock('@/stores/toastStore', () => ({
  ToastStore: jest.fn()
}));

describe('ToastContainer Component', () => {
  const mockHideToast = jest.fn();

  // Default mock values for ToastStore
  const defaultStoreMock = {
    message: 'Test Toast Message',
    type: 'success',
    isVisible: true,
    undoEnabled: false,
    timeoutDuration: 3000,
    hideToast: mockHideToast
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (ToastStore as jest.MockedFunction<typeof ToastStore>).mockReturnValue(defaultStoreMock);
  });

  it('should match the snapshot when toast is visible', () => {
    const { container } = render(<ToastContainer />);
    expect(container).toMatchSnapshot();
  });

  it('should render the toast message when visible', () => {
    render(<ToastContainer />);
    expect(screen.getByText('Test Toast Message')).toBeInTheDocument();
  });

  it('should not render the toast when isVisible is false', () => {
    (ToastStore as jest.MockedFunction<typeof ToastStore>).mockReturnValue({
      ...defaultStoreMock,
      isVisible: false
    });

    render(<ToastContainer />);
    expect(screen.queryByText('Test Toast Message')).not.toBeInTheDocument();
  });

  it('should hide the toast after the timeout duration', () => {
    jest.useFakeTimers();
    render(<ToastContainer />);

    jest.advanceTimersByTime(3000);
    expect(mockHideToast).toHaveBeenCalledTimes(1);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
