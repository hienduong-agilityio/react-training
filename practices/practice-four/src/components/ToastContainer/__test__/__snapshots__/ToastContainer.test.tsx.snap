// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastContainer Component should match the snapshot when toast is visible 1`] = `
<div>
  <div
    class="fixed top-9 left-1/2 transform -translate-x-1/2"
  >
    <div
      class="flex items-center divide-x rtl:divide-x-reverse p-4 w-full max-w-xs rounded-lg shadow text-success-100 bg-success-800 divide-green-950"
    >
      <div
        class="text-sm font-normal inline-flex items-center gap-3 pl-1"
      >
        <img
          alt="success"
        />
        Test Toast Message
      </div>
    </div>
  </div>
</div>
`;
